Draft : Dark light mode

lealabou opened the pull request:
**Début d'ajout du plugin dark/light/both mode.**

Cette pull request a pour but d'avoir un retour sur mon code concernant l’intégration d’un plugin de changement de thème.

Description:
- Le code principal se trouve dans src/ts/plugins/grapejs-theme-mode.ts.
- Pour l'instant le code affiche juste un selecteur dans le menu Style Manager de Silex
- Le selecteur ne récupère pas encore les éléments css modifiés.

lexoyo commented on package.json:
don't commit these changes pls

but it would be awesome to move these changes to another another PR/feature "WIP: make silex run on windows" as we will need this

lexoyo commented on src/css/_variables.scss:
don't commit this pls

lexoyo commented on src/html/index.pug:
i think this is not related to dark mode in the websites but the editor's dark mode?

lealabou replied:
No this is not related, it will be deleted on the next commit

lexoyo commented on src/ts/plugins/grapesjs-theme-mode.ts:
move this file to `/src/ts/client/grapesjs/` please

lexoyo replied:
The feature should be in a separate file, as it is a grapesjs plugin

lexoyo commented on src/ts/plugins/grapesjs-theme-mode.ts:
```diff
@@ -0,0 +1,105 @@
+  //console.log('Plugin grapesjs-theme-mode chargé !');
+
+  //define the current mode to both
+  (editor as any).ThemeMode = { current: 'both' }
```
You don't need to add a property to editor in this case

The best would be to not keep the state anywhere, but if you really need to, keep the state in a global variable

You can also fire events on `editor` if you need other parts to know about this state

lexoyo commented on src/ts/plugins/grapesjs-theme-mode.ts:
```diff
@@ -0,0 +1,105 @@
+  (editor as any).ThemeMode = { current: 'both' }
+
+  //add the sector to the StyleManager
+  const injectThemeSector = () => {
```
use `function injectThemeSector {` instead

lexoyo commented on src/ts/plugins/grapesjs-theme-mode.ts:
```diff
@@ -0,0 +1,105 @@
+      return
+    }
+
+    sm.addSector('theme-mode', {
```
Don't add a themeMode selector in the Style Manager. It doesn't belong there.
You want your selector to be somewhere in the top bar, like the responsive / device / breakpoint selector

Here:
![image](https://github.com/user-attachments/assets/a97382c4-bff5-4b18-b1ef-54dab57d08d3)

Or maybe ask the community about the UX

But for sure this doesn't belong to the "css styles" part

lealabou replied:
Yes we talk about it with Olivier, I'm going to move it besides the Desktop/Tablet/mobile manager.

lealabou replied:
Done

lexoyo reviewed (commented):
I believe you need to start a discussion about the UI, maybe with screenshots of what it will be like

lealabou commented:
What is added in the new changes:

- Modified the branch and is now up to date with no irrelevent file
- Added addThemeSelector function to create a theme selector dropdown. (will be added in a grapesJS pluggin file in the next commit.
- Implemented saveCurrentStylesToTheme to save styleManager properties to themeStyles for the selected component.
Implemented updateStyleManagerForTheme to apply saved styles to the StyleManager when switching themes or selecting a component.
- Saved theme mode in the components themeMode property and styles in themeStyles.
- Added updateComponentThemeStyles only working on the id for now not for classes (will be added in my next commit), for adding css rules and media queries.
- There is some issues that will be fixed in my next commit.

lexoyo commented on src/ts/client/grapesjs/index.ts:
```diff
@@ -437,13 +438,145 @@ export async function initEditor(config: EditorConfig) {
+function addThemeSelector(editor) {
+  setTimeout(() => {
+    // Add a selector besides the deviceManager
+    const viewsBar = document.querySelector('.gjs-pn-panel.gjs-pn-devices-c')
```
I believe there is a grapesjs api to get a reference to the option panel, can you check if it can be used here?

lealabou replied:
I tried to used getPannel but it doesn't work, i've added it in the new commit

lexoyo commented on src/ts/client/grapesjs/index.ts:
```diff
@@ -437,13 +438,145 @@ export async function initEditor(config: EditorConfig) {
+        if (!model) return
+        const theme = model.get('themeMode') || 'both'
+        lastTheme = theme
+        const select = document.getElementById('gjs-theme-select')
```
Don't retrieve the select every time something changes

lealabou replied:
done

lexoyo commented on src/ts/client/grapesjs/index.ts:
```diff
@@ -437,13 +438,145 @@ export async function initEditor(config: EditorConfig) {
+      // Listen to the component selection and update the selector
+      editor.on('component:selected', (model) => {
+        if (!model) return
+        const theme = model.get('themeMode') || 'both'
```
Why store the theme on the components ?

lealabou replied:
I've put it more globally now, for me before it was logical to put it on the component but when you said it I saw the issue and I changed it

lealabou commented:
What is added in the new changes:

- fixes on bug like the css properties that was shared between element
- fixes on previous note on the PR

I have two main issues about the code that I pushed:
- I did a code for the classes but I didn't know why it wasn't working I figured out that when I try to get the classes it put undefined. (see in the new function getComponentClassNames that try to have the class name on a component). 
- I don't know hot to put the theme selector besides the devices-c panel by using that: panelManager.getPanel('devices-c') because it doesn't work.

lealabou commented:
It's working now for the classes it will be added in my next commit, i just have to clean the code and I send it tomorrow morning

lexoyo commented on src/ts/client/grapesjs/index.ts:
```diff
@@ -368,7 +367,7 @@ export async function initEditor(config: EditorConfig) {
       reject(e)
     }
-
+    
```
useless spaces

lexoyo commented on src/ts/client/grapesjs/index.ts:
```diff
@@ -437,13 +439,165 @@ export async function initEditor(config: EditorConfig) {
   })
 }
 
+// Will be added later in a pluggin but for the test I put it here
```
yes pls

lealabou replied:
done

lexoyo commented on src/ts/client/grapesjs/index.ts:
```diff
@@ -437,13 +439,165 @@ export async function initEditor(config: EditorConfig) {
+  let themeSelect = null 
+  setTimeout(() => {
+    const panelManager = editor.Panels
+    // problem here, the pannel devices-c is found but has no 'el' property so it's the document.querySelector that is used
```
do not use `.get`, inspect what's returned by `getPanel`

lexoyo commented on src/ts/client/grapesjs/index.ts:
```diff
@@ -437,13 +439,165 @@ export async function initEditor(config: EditorConfig) {
+
+  // Save all current style values to the given theme
+  function saveCurrentStylesToTheme(model, theme) {
+    const themeStyles = model.get('themeStyles') || { both: {}, light: {}, dark: {} }
```
I don't think it's the way to go to store the styles at the component level
Why would you do that? The styles are stored in the StyleManager already... I think
Does it work?

lealabou replied:
Yes it works, I did it that way because It permit to stock multiple theme to a component.

lexoyo commented on src/ts/client/grapesjs/index.ts:
```diff
@@ -437,13 +439,165 @@ export async function initEditor(config: EditorConfig) {
+        const value = e.target.value
+        const selected = editor.getSelected()
+        if (selected) {
+          saveCurrentStylesToTheme(selected, currentTheme)
```
why store the styles to the previous theme properties?

lealabou replied:
I saved styles to the current theme before updating currentTheme to ensure all modifications are stored for the active theme

lealabou commented:
What the new commit does:
- Implemented theme support for classes
- move the function to a new file
- Tested in Editor and Published site and it works.

I will implement the advices that you have given in my next commit !

lealabou commented:
I didn't succeed to fix any of theses problem that I present below.

Things that still need to be fixed about this feature:

- Correcting the CSS that duplicates sometimes when we select an element that already has a CSS it duplicates into the one that was previously selected.
- Utilize the getPanel to have access to the devices-c instead of the document.querySelector
- Add features like switch buttons or selectors for the users to change the mode of the published website instead of using the preferences of the navigator/computer.
